File: 프로그래머스/2/154538. 숫자 변환하기/숫자 변환하기.py
def solution(x, y, n):
    
    if x == y:
        return 0
    
    queue = [(x, 0)]
    visited = set()
    
    while queue:
        current_value, count = queue.pop(0)
        for next_value in (current_value * 2, current_value * 3, current_value + n):
            if next_value == y:
                return count + 1
            
            if next_value < y and next_value not in visited:
                visited.add(next_value)
                queue.append((next_value, count + 1))
    
    return -1

File: 프로그래머스/2/154538. 숫자 변환하기/숫자 변환하기.java
import java.util.*;

class Solution {
    public int solution(int x, int y, int n) {
        Queue<int[]> queue = new LinkedList<>();
        boolean[] visited = new boolean[y+1];
        
        queue.offer(new int[]{x, 0});
        visited[x] = true;
        
        while (!queue.isEmpty()) {
            int[] now = queue.poll();
            int value = now[0];
            int count = now[1];
            
            if (value == y)
                return count;
            int[] nextValue = {value + n, value * 2, value * 3};
            for (int next : nextValue) {
                if (next <= y && !visited[next]) {
                    visited[next] = true;
                    queue.offer(new int[]{next, count + 1});
                }
            }
        }
        
        return -1;
    }
}